policies/recipes/artifacts_test.go

```diff
@@ -0,0 +1,302 @@
+		t.Fatalf("could not clean up test data")
+	}
+
+	if art["id1"] != "/tmp/id1.deb" {
+		t.Errorf("expected file entry, but not present")
+	}
+}
+
+func TestFetchArtifacts_http_InvalidURL(t *testing.T) {
+	// Start a local HTTP server
+	server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		// Test request parameters
+		rw.Write([]byte("ok"))
+		// Send response to be tested
+	}))
+	// Close the server when test finishes
+	defer server.Close()
+
+	getHTTPClient = func() (*http.Client, error) {
+		return server.Client(), nil
+	}
+	remoteartifact := osconfigpb.SoftwareRecipe_Artifact_Remote{
+		Uri:      fmt.Sprintf("%sagent.deb", server.URL),
+		Checksum: "",
+	}
+	artifacts := []*osconfigpb.SoftwareRecipe_Artifact{
+		&osconfigpb.SoftwareRecipe_Artifact{
+			Id:            "id1",
+			Artifact:      &osconfigpb.SoftwareRecipe_Artifact_Remote_{Remote: &remoteartifact},
+			AllowInsecure: true,
+		},
+	}
+
+	art, err := FetchArtifacts(context.Background(), artifacts, "/tmp/")
+
+	if len(art) != 0 {
+		t.Errorf("unexpected artifact downloaded")
+	}
+
+	if err == nil || !strings.Contains(err.Error(), "Could not parse url") {
+		t.Errorf("expected url error, found(%+v)", err)
+	}
+
+}
+
+func TestFetchArtifacts_http_NonHTTPRemote(t *testing.T) {
+	// Start a local HTTP server
+	server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		// Test request parameters
+		rw.Write([]byte("ok"))
+		// Send response to be tested
+	}))
+	// Close the server when test finishes
+	defer server.Close()
+
+	getHTTPClient = func() (*http.Client, error) {
+		return server.Client(), nil
+	}
+
+	url := strings.Replace(server.URL, "http", "ftp", -1)
+	remoteartifact := osconfigpb.SoftwareRecipe_Artifact_Remote{
+		Uri:      fmt.Sprintf("%s/agent.deb", url),
+		Checksum: "",
+	}
+	artifacts := []*osconfigpb.SoftwareRecipe_Artifact{
+		&osconfigpb.SoftwareRecipe_Artifact{
+			Id:            "id1",
+			Artifact:      &osconfigpb.SoftwareRecipe_Artifact_Remote_{Remote: &remoteartifact},
+			AllowInsecure: true,
+		},
+	}
+
+	art, err := FetchArtifacts(context.Background(), artifacts, "/tmp/")
+
+	if len(art) != 0 {
+		t.Errorf("unexpected artifact downloaded")
+	}
+
+	if err == nil || !strings.Contains(err.Error(), "unsupported protocol") {
+		t.Errorf("expected url error, found(%+v)", err)
+	}
+
+}
+
+func TestFetchArtifacts_http_FetchError(t *testing.T) {
+	// Start a local HTTP server
+	server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		// Test request parameters
+		rw.Write([]byte("ok"))
+		// Send response to be tested
+	}))
+	// Close the server when test finishes
+	defer server.Close()
+
+	getHTTPClient = func() (*http.Client, error) {
+		return server.Client(), nil
+	}
+
+	remoteartifact := osconfigpb.SoftwareRecipe_Artifact_Remote{
+		Uri:      fmt.Sprintf("http://non-existent-domain.com/agent.deb"),
```

Comment:
what is the behavior of httptest.Client here? will it recognize this is not served by the httptest server and fail quickly?

Reply:
right, since the server.client() returns a client that is configured to talk to the endpoint of "server", the client will fail to connect to "non-existent-domain.com" and will fail.

anyhow, this test is also removed